common/djangoapps/student/tests/test_models.py

```diff
@@ -713,6 +714,114 @@ def test_not_enrolled_student_is_enrolled(self):
+        )
+
+        # # trigger the post_save callback
+        student.is_active = True
+        student.save()
+
+        actual_cea = CourseEnrollmentAllowed.objects.get(email=student.email)
+
+        assert actual_cea.user is None
+
+    @mock.patch.dict('django.conf.settings.FEATURES', {'DISABLE_AUTOENROLL_AFTER_COURSE_END': False})
+    def test_student_enrolled_if_course_ended_and_flag_false(self):
+        """
+        Make sure that invited students who are not enrolled gets enrolled
+        when their account is activated even if the course has ended.
+        This is the default behavior without
+        DISABLE_AUTOENROLL_AFTER_COURSE_END flag
+        """
+
+        # Set end date to a past date so that course is ended
+        start_date = datetime.datetime.now(pytz.UTC) - datetime.timedelta(days=3)
+        end_date = datetime.datetime.now(pytz.UTC) - datetime.timedelta(days=2)
+        course_overview = CourseOverviewFactory.create(
+            id=self.course.id,
+            start=start_date,
+            end=end_date,
+        )
+        course_overview.save()
+
+        # fixture
+        student = self._set_up_invited_student(
+            course=self.course,
+            active=False,
+            enrolled=False
+        )
+
+        # # trigger the post_save callback
+        student.is_active = True
+        student.save()
+
+        actual_cea = CourseEnrollmentAllowed.objects.get(email=student.email)
+
+        assert actual_cea.user == student
+
+    @mock.patch.dict('django.conf.settings.FEATURES', {'DISABLE_AUTOENROLL_AFTER_COURSE_END': True})
+    def test_student_enrolled_if_course_restarted_and_flag_true(self):
+        """
+        Make sure that invited students who are not auto-enrolled
+        during registration as the course had ended, gets auto-enrolled
+        during login if the course reopens due to extension of course end
+        date.
+        """
+
+        # Set end date to a past date so that course is ended
+        start_date = datetime.datetime.now(pytz.UTC) - datetime.timedelta(days=3)
+        end_date = datetime.datetime.now(pytz.UTC) - datetime.timedelta(days=2)
+        course_overview = CourseOverviewFactory.create(
+            id=self.course.id,
+            start=start_date,
+            end=end_date,
+        )
+        course_overview.save()
+
+        # fixture
+        student = self._set_up_invited_student(
+            course=self.course,
+            active=False,
+            enrolled=False
+        )
+
+        # # trigger the post_save callback
+        student.is_active = True
+        student.save()
+
+        end_date = datetime.datetime.now(pytz.UTC) + datetime.timedelta(days=2)
+        course_overview.end = end_date
+        course_overview.save()
+
+        # student.last_login = datetime.datetime.now(pytz.UTC)
+        student.save()
+
+        actual_cea = CourseEnrollmentAllowed.objects.get(email=student.email)
+
+        assert actual_cea.user == student
```

Comment:
That's a lot of code repetition. Could we simplify it with `@ddt.data`?

Reply:
Done